src/lib/upgradePackageDefinitions.ts

```diff
@@ -49,18 +100,42 @@ export async function upgradePackageDefinitions(
 
   if (options.peer && Object.keys(filteredLatestDependencies).length > 0) {
     const upgradedPeerDependencies = await getPeerDependenciesFromRegistry(filteredLatestDependencies, options)
-    const peerDependencies = { ...options.peerDependencies, ...upgradedPeerDependencies }
-    if (!dequal(options.peerDependencies, peerDependencies)) {
-      const [newUpgradedDependencies, newLatestVersions, newPeerDependencies] = await upgradePackageDefinitions(
-        { ...currentDependencies, ...filteredUpgradedDependencies },
-        { ...options, peerDependencies, loglevel: 'silent' },
-      )
-      return [
-        { ...filteredUpgradedDependencies, ...newUpgradedDependencies },
-        { ...latestVersionResults, ...newLatestVersions },
-        newPeerDependencies,
-      ]
+
+    let checkPeerViolationResult: CheckIfInPeerViolationResult = {
+      violated: false,
+      filteredUpgradedDependencies,
+      upgradedPeerDependencies,
     }
+    let rerunResult: UpgradePackageDefinitionsResult
+    let runIndex = 0
+    do {
```

Comment:
I'm struggling to understand this do-while loop. Something about the `runIndex > 1` "base case" and the constant reassignment of `rerunResult. 

Any ideas how to clean this up and make it easier to understand?

Reply:
I gave it another shot